### PR TITLE
Use messagepack

### PR DESCRIPTION
**Pull request by jchannon**

### CODE REVIEW

**Review by ritasker — APPROVED** [review]

File: src/Sinks/FluentdSink.cs [hunk 1]
```diff
@@ -160,73 +130,52 @@ private void Format(LogEvent logEvent, StringWriter output)
                     name = '@' + name;
                 }
 
-                output.Write(',');
-                JsonValueFormatter.WriteQuotedJsonString(name, output);
-                output.Write(':');
-                new JsonValueFormatter("$type").Format(property.Value, output);
+                ((IDictionary<string, object>)localEvent).Add(name, "");
+                var visitor = new SerilogVisitor();
```
Comment (jchannon) [inline src/Sinks/FluentdSink.cs]: Wonder if we should create one instance of the visitor in the ctor?

File: src/Sinks/FluentdSink.cs [hunk 2]
```diff
@@ -160,73 +130,52 @@ private void Format(LogEvent logEvent, StringWriter output)
...
+                visitor.Visit((IDictionary<string, object>)localEvent, name, property.Value);
             }
 
-            output.Write('}');
+            var serializationContext = new SerializationContext(PackerCompatibilityOptions.PackBinaryAsRaw) { SerializationMethod = SerializationMethod.Map };
```
Comment (jchannon) [inline src/Sinks/FluentdSink.cs]: Wonder if we should create one instance of the serialization context

File: src/Sinks/FluentdSink.cs (hunk 1 above)
Comment (JoeStead) [reply src/Sinks/FluentdSink.cs]: Yes

File: src/Sinks/FluentdSink.cs (hunk 2 above)
Comment (JoeStead) [reply src/Sinks/FluentdSink.cs]: Yes

File: src/Sinks/SerilogVisitor.cs [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    using System;
+    using System.Collections.Generic;
+    using Serilog.Events;
+
+    public class SerilogVisitor
+    {
+        public void Visit(IDictionary<string, object> state, string name, LogEventPropertyValue value)
+        {
+            var returnValue = this.Visit(value);
+            state[name] = returnValue;
```
Comment (JoeStead) [inline src/Sinks/SerilogVisitor.cs:12]: Whoever came up with this is bloody brilliant ;)

**Review by JoeStead — APPROVED** [review]

File: src/Sinks/FluentdSink.cs [hunk 4]
```diff
@@ -136,8 +136,6 @@ private void FormatwithMsgPack(LogEvent logEvent, Packer packer)
                     name = '@' + name;
                 }
 
-                ((IDictionary<string, object>)localEvent).Add(name, "");
```
Comment (JoeStead) [inline src/Sinks/FluentdSink.cs]: I wondered about this :)

**Review by JoeStead — APPROVED** [review]

File: src/Sinks/SerilogVisitor.cs (hunk 3 above)
Comment (jchannon) [reply src/Sinks/SerilogVisitor.cs:12]: Thanks

**Pull request merged**